### PR TITLE
add tensor type in Go

### PR DESCRIPTION
**Pull request by QiJune**

### CODE REVIEW

File: elasticdl/pkg/common/tensor.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+// TODO(qijun): handle different tensor dtype
+type Tensor struct {
+	Name    string
+	Value   []float32
+	Dim     []int64
+	Indices []int64
+}
+
+// GetTensorSize returns the size of a tensor
+func GetTensorSize(t *Tensor) int64 {
```
Comment (skydoorkai) [inline elasticdl/pkg/common/tensor.go]: We can use method of struct instead:
```
func (t *Tensor)Size() int64 {
   ...
}
```
Then for a `tensor`, call `tensor.Size()` to get its size.

File: elasticdl/pkg/common/tensor.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+	var size int64 = 1
+	for _, d := range t.Dim {
+		size *= d
+	}
+	return size
+}
+
+// DeserializeTensorPB transforms pb to tensor
+func DeserializeTensorPB(pb *proto.Tensor, t *Tensor) {
```
Comment (skydoorkai) [inline elasticdl/pkg/common/tensor.go]: What is the usage of this func? Should `Tensor` be pre-defined? Or we can define a func with a input as `*proto.Tensor` and returns a `*Tensor`?

File: elasticdl/pkg/common/tensor.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+	copy(t.Dim, pb.Dim)
+	t.Indices = make([]int64, len(pb.Indices))
+	copy(t.Indices, pb.Indices)
+	t.Value = make([]float32, len(pb.Content)/4)
+	br := bytes.NewReader(pb.GetContent())
+	binary.Read(br, binary.LittleEndian, &t.Value)
+}
+
+// SerializeTensor transforms tensor to pb
+func SerializeTensor(t *Tensor, pb *proto.Tensor) {
```
Comment (skydoorkai) [inline elasticdl/pkg/common/tensor.go]: Same for this method.

Comment (QiJune) [reply elasticdl/pkg/common/tensor.go]: Done

File: elasticdl/pkg/common/tensor.go (hunk 2 above)
Comment (QiJune) [reply elasticdl/pkg/common/tensor.go]: Modify this function, to get a `*proto.Tensor`, and return a `*Tensor`

File: elasticdl/pkg/common/tensor.go (hunk 1 above)
Comment (QiJune) [reply elasticdl/pkg/common/tensor.go]: Rename It to `GetDimProduct`, so it could be used for any slice.

File: elasticdl/pkg/common/tensor.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+// TODO(qijun): handle different tensor dtype
+type Tensor struct {
+	Name    string
+	Value   []float32
+	Dim     []int64
+	Indices []int64
+}
+
+// GetDimProduct returns the product of a slice of dim
+func GetDimProduct(dim []int64) int64 {
```
Comment (skydoorkai) [inline elasticdl/pkg/common/tensor.go]: Will `GetDimProduct` be used outside this file? If not, change it to an unexported func. If yes, keep as it is.

Comment (QiJune) [reply elasticdl/pkg/common/tensor.go]: I am not very sure. 
We could set it to an unexported func first. If it is used in another place, we change it.

**Review by skydoorkai — APPROVED** [review]

**Pull request merged**